### ISSUE TITLE
coarser testing predicates

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#1378]
The opposite of `is` is not `is-not`.

Specifically, suppose I have test T and I expect it to pass on implementation I: I can write `I(T) is …`. 

But suppose we're in an adversarial setting: one person writes T and it's being run against someone else's I's. Furthermore, T is written to fail. It's impossible to predict how the unseen I is going to work: it may successfully complete and produce a different answer, in which case `I(T) is-not …` makes sense, or T may be so fiendish relative to I that it errors: `I(T) raises …`. 

The author of the test suite can't predict which will happen, and hence what should be a testing success looks like a testing failure. It would be nice to have a predicate that is really the opposite of `is` that represents "anything but `is`".

NB: @jswrenn has run into this in practice when trying to assess the quality of oracular testing. This posting is on his behalf.

**Comment by shriram** [comment]
Alternatively, we can think of `is-not` having this broader interpretation in the first place. "I don't care how it fails to not be this, I just want it to not be this." (Consequences for other `-not`s not thought through.)

**Comment by blerner** [comment]
The full set of negative tests are:
```
is-not, is-not==, is-not=~, is-not<=>
violates
raises-other-than
raises-violates
```
I think it's pretty clear that `raises-other-than` and `raises-violates` should not be considered to pass if the code under test does not even raise an error.  I think the `is-not*` forms are uniform in their treatment: if it makes sense for `is-not`, it'll make sense for the others as well.  I'm not sure on `violates`.

So, we don't currently have a form for "does anything except return this one answer" and "does anything except raise this one exception".

In the Pyret test suite, we use `run-task` for functions that may fail: it returns a Result that we can then pattern match on.  It's not as easy to use as the normal testing forms, but it's like an analogue of equal3 versus equal -- you get more information and more stability, at the price of more explicit code.  Here's a complete example:

```
include either
include error
include srcloc

fun does-not-produce(expected):
  lam(thunk):
    ans = run-task(thunk)
    spy: 
      ans: ans,
      contents: cases(Either) ans:
        | left(actual) => actual
        | right(err) => exn-unwrap(err)
      end
    end
    cases(Either) ans:
      | left(actual) => actual <> expected
      | right(_) => true
    end
  end
end

examples:
  lam(): raise(no-cases-matched(builtin("dummy"), {rich: true, value: false})) end satisfies does-not-produce(5)
  lam(): 1 / 0 end satisfies does-not-produce(5)
  lam(): 5 / 1 end violates does-not-produce(5)
end
```

**Comment by shriram** [comment]
It's funny, when I started to read your message, the phrase that came to mind as a contrast to `is-not` (with its current semantics retained) was `does-not`, but I thought, "but that doesn't parse, it really ought to be `does-not-produce`". And then I read to the end…

**Comment by shriram** [comment]
One difficulty is, students are still going to be writing `is-not` in their tests, because they are not very good at thinking about adversarial behavior. [*] 

So would we need to grep and replace every `is-not` with `does-not-produce`? That seems weird. It seems better to flip the meanings: the default is forgiving, but if you really want to check more selectively (this had better produce a value _and_ it had better be different), then you specifically write that in, and then your intent is clear, and a violation of that is a genuine violation.

[*] In 019 placement, students are genuinely confused about how they can turn in a test suite without an implementation, and how this could be run against someone else's code — which is a much lower-level leap than this distinction. There's some conceptual gap there.

**Comment by blerner** [comment]
I misunderstood your original post to say that @jswrenn was trying to run student tests and needed more generic predicates, not that students were trying to write tests that were unaware of the distinction between not evaluating and not raising.  I'm slightly more sympathetic to the need for a new testing operator, then.

I worry about bikeshedding here, specifically `does-not-produce` (unary) vs `does-not-raise` (0-ary) versus `raises-other-than` (unary but different than `does-not-produce`).  I want test operators that sound similar to behave similarly.  And I worry about how this interacts with `because` -- should something like `1 / 0 does-not-produce 5 because 3` succeed?  (3 doesn't evaluate to 5, but 1/ 0 does not evaluate at all...)

**Comment by jswrenn** [comment]
When @shriram said I had actually run into this issue, it wasn't that I had encountered students mis-using `is-not`, it was that an oracle doesn't just signal failure by _only_ by returning `false`; it can also raise an exception. The negative test `oracle(really-bad-impl) is false` doesn't cut it in that scenario, but `is-not true` _also_ doesn't work; both look like a "failing" test to Pyret, even though some degree of failure was expected.

So far, we've discussed two possible changes we can make:
 1. Keep the current meaning of `is-not`, so that `raise(...) is-not true` _fails_, and add a separate, more general testing operator.
 2. Generalize `is-not` so that `raise(...) is-not true` _passes_, and add a separate, more specific operator that distinguishes between the value case.

Intuitively: I'd like to see `raise(...) is-not true` _pass_, because, read aloud, that statement _sounds_ true, and then add some other operator that _only_ accepts the value case.

But: If students grasp negative testing before they grasp the range of situations in which exceptions can occur (pyret-based _and_ user-based), then keeping the current meaning of `is-not` makes the most sense, since exceptions are still weird and unexpected things. Remember, there's all sorts of useful debugging information you get when a test fails, but **not** when a test passes!

In short: I'm not sure which is best. I think generalizing `is-not` is the simplest and most intuitive option, but it could hide _unexpected_ exceptions from students. As of now, I don't have any strong reasons to favor either option.

---
Addendums:

1. @shriram pointed out that some very interesting uses of `is-not` that I had not considered, and that we should grep through the assignments to see whether students are even using the negative testing operators. I'll do so.
2. I'm curious whether the issues I encountered would go away if students did oracle in a statically typed context. We should consider this.

**Comment by sorawee** [comment]
`raise` has bottom type which is a subtype of everything. I don't think Pyret type system will help here.

**Comment by blerner** [comment]
> The author of the test suite can't predict which will happen, and hence what should be a testing 
> success looks like a testing failure. It would be nice to have a predicate that is really the opposite of is 
> that represents "anything but `is`".

`is-anything-but`?

I'm torn on whether I want the semantics of `is-not` to be this forgiving: I don't think students writing tests for their _own_ code should be insulated from unintentional exceptions in their own code.  E.g. if we were giving a first-week exercise to write a `reciprocal` function, the most important point of that function is what it does for an argument of 0.  Saying merely that "well, whatever it does, it's not zero!" seems like a cop-out test...
